sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

```diff
@@ -0,0 +1,373 @@
+We expect implementations of this ABI to build on Clang's `ext_vector_type`,
+rather than specifying the register types directly.
+
+The value of `N` for each `simd` argument to a SIMD function must be the same,
+and represents the sub-group size of the calling kernel.  If a function accepts
+`simd` arguments with multiple `N` values, it cannot be invoked from a SPMD
+kernel and the implementation must issue a diagnostic in this case.
+
+The example below shows a simple SIMD function that scales all elements of
+an 8-wide SIMD type `x` by a scalar value `n`.
+
+[source, c++]
+----
+sycl::ext::oneapi::experimental::simd<float, 8> scale(sycl::ext::oneapi::experimental::simd<float, 8> x, float n)
+{
+  return x * n;
+}
+----
+
+=== Invoking a SIMD Function
+
+The `invoke_simd` free function invokes a SIMD function using all work-items
+in a `sub_group`.  The `invoke_simd` interface marshals data between the
+SPMD context of the calling kernel and the SIMD context of the callee,
+converting arguments and return values between scalar and SIMD types as
+appropriate.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+template <typename Callable, typename...T>
+auto invoke_simd(sub_group sg, Callable&& f, T... args);
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+The `Callable` passed to `invoke_simd` must be a C++ callable that can be
+invoked with the same number of arguments specified in the `args` parameter
+pack. `Callable` may be a function object, a lambda,
+or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
+`Callable` must be an immutable callable with the same type and state for all
+work-items in the sub-group, otherwise behavior is undefined.
+
+NOTE: When using a lambda `Callable`, it is very easy for a user to
+unintentionally capture non-uniform variables and violate the requirements
+from the previous paragraph. This form of `invoke_simd` should be used with
+caution.
+
+Each argument to the invoked `Callable` must be a scalar type, a SIMD type,
+or a `std::tuple` of these types. The length of each SIMD type must match the
+sub-group size of the kernel; calling a SIMD function with mismatched lengths
+is an error, and an implementation must issue a diagnostic.
+
+To enable overloading in cases where the `Callable` accepts either a scalar
+type or a SIMD type, we leverage the `sycl::ext::oneapi::experimental::uniform` wrapper
+type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.
+
+Each argument in the `args` parameter pack must be an arithmetic type, a
+trivially copyable type wrapped in a `sycl::ext::oneapi::experimental::uniform`, or a
+`std::tuple` of these types.  Arguments may not be pointers or references,
+but pointers (like any other non-arithmetic type) may be passed if wrapped in a
+`sycl::ext::oneapi::experimental::uniform`.  Any such pointer value must point to memory
```

Comment:
This basically forbids to have a simd callee with vector-of-pointers argument:
```cpp
void simd_foo(simd<int*,16> ptrs, ...) { scatter(ptrs,...); }
```
And the spec implies non-uniform pointer is represented as a base pointer + a vector of offsets, so SIMD library functions must be written with this style to be callable from SYCL:
```cpp
void simd_foo(int *base, simd<uintptr_t,16> offsets, ...) { ... }
```
This is natural for many platforms, but this implication should better be explicitly mentioned, I think.

(this and my other late comments can be addressed in future revisions of the spec, not a blocker for merge)

Reply:
I don't think this is a significant usability issue, because you can write a wrapper:
```c++
void simd_foo_from_sycl(int *base, simd<uintptr_t,16> offsets, ...) {
  simd<int*, 16> ptrs(base + offsets);
  simd_foo(ptrs, ...);
}
```

Would you prefer a non-normative note pointing out this is possible, or an unresolved issue suggesting we should revisit this restriction?

Reply:
Ah, you are right, thanks. As simple as that :)

> Would you prefer a non-normative note pointing out this is possible, or an unresolved issue suggesting we should revisit this restriction?

Non-normative note would be enough.

Reply:
Added in https://github.com/intel/llvm/pull/3445/commits/b91567f38984c49f529b0ab2dc215fce83563f89.